### PR TITLE
rework on ads client 

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
This pr mainly addresses two issues:

1. grpc client will reconnect, reply on it and so no connection leak

2.  auto rotate tls certificate

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
Has to work on stream cleanup

**Review by noumyxxo — APPROVED** [review]

**Review by noumyxxo — APPROVED** [review]

**Review by noumyxxo — APPROVED** [review]

**Review by noumyxxo — APPROVED** [review]

**Review by esnible — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
@costinm Please take a look when you have bandwidth

File: pkg/adsc/adsc.go [hunk 1]
```diff
@@ -305,42 +315,47 @@ func (a *ADSC) tlsConfig() (*tls.Config, error) {
...
-		key, err := a.cfg.Secrets.GenerateSecret(context.Background(), "agent",
-			cache.WorkloadKeyCertResourceName, string(tok))
-		if err != nil {
-			return nil, err
-		}
-		clientCert, err = tls.X509KeyPair(key.CertificateChain, key.PrivateKey)
-		if err != nil {
-			return nil, err
+	if a.cfg.SecretManager != nil {
+		getClientCertificate = func(requestInfo *tls.CertificateRequestInfo) (*tls.Certificate, error) {
```
Comment (costinm) [inline pkg/adsc/adsc.go:321]: Can you just move this to a proper method (getClientCert)  ? not a big deal, can be done in separate PR

**Review by costinm — APPROVED** [review]
Looks good, thanks.

File: pkg/adsc/adsc.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/adsc/adsc.go:321]: ok

**Pull request merged**